### PR TITLE
Add sliding window

### PR DESCRIPTION
**Pull request by gkielian**
Adds option to utilize a sliding window in place of the standard attention.

Setting default for sliding window to be False.

### CODE REVIEW

**Comment by msaligane** [comment]
@777yifei @advancedengineering can you please review.

**Comment by advancedengineering** [comment]
I am reading sliding window now, I might misunderstand the code but shouldn't the mask for the window be a span of W(hyper parameter, window size) around the diagonal of the matrix?
In my understanding, self.mask is a lower triangular matrix, and after cropping, it will still be a lower triangular matrix like the vanilla attention not sliding window.
mask = self.mask[:, :, -T:, -T:] # crop mask to sequence length

File: data/modular_addition/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+```
+ 0123456789abcef
+
+```
+
+Above would have space, and newlines as characters in addition to base 16 chars.
+
+3. Process the examples into a training set:
+```bash
+python3 prepare.py --input_file data/base_16.txt --token_file tokens.txt
```
Comment (777yifei) [inline data/modular_addition/README.md:28]: Readme file is really clear, but I am a little bit confused about whta is base_16.txt here

Comment (777yifei) [reply data/modular_addition/README.md:28]: Hi Greg, does the token_file flag let tokens.txt overwrites base_16.txt?

File: data/modular_addition/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+These files can then be used to train modular arithmetic models.
+
+## Creating Tokenisation Straight From Text Files
+
+If the `token_file` flag is omitted from the invocation, then the tokenization
+is drawn from the `input_file`.
+
+For example:
+```bash
+python3 prepare.py --input_file data/base_2.txt
```
Comment (777yifei) [inline data/modular_addition/README.md:42]: I am not sure about what is base2.txt here

**Review by 777yifei — COMMENTED** [review]
Pertty clear Readme, easily to understand function of each file. But there are still few confustion

File: model.py [hunk 3]
```diff
@@ -144,6 +160,7 @@ class GPTConfig:
     n_embd: int = 768
     dropout: float = 0.0
     bias: bool = True # True: bias in Linears and LayerNorms, like GPT-2. False: a bit better and faster
+    use_sliding_window: bool = False # True: uses softermax; False uses softmax
```
Comment (777yifei) [inline model.py:163]: There is a comment issue

**Review by 777yifei — CHANGES_REQUESTED** [review]

File: data/modular_addition/README.md (hunk 1 above)
Comment (gkielian) [reply data/modular_addition/README.md:28]: Thanks for the question! 

Ah looks like I have to rebase this PR, the modular addition folder is separate from the sliding window.

The modular arithmetic folder is part of the parameter count reduction effort, in particular tokenization optimization for reducing total parameter count.

Will share images soon showing the training and validation losses for each of these (just posted a tensorboard integration PR yesterday).

Essentially, there are demonstrable costs (more things to learn) for certain tokenization starting points:

1. synthesis costs for too fine a tokenization -- e.g.  letters, or pen strokes, for written language might require a lot of synthesis to get to meaningful phoneme content (which can then be used for rhyming and meaning)
2. decomposition costs for too coarse a selection -- e.g. english word based tokenization, where the phonemeand syllable locations need to be "discovered" through many examples.

So one theory we can write about is that choosing a phoneme tokenization removes these costs and results in the resulting lower perplexity Shiwei observed.

(These are also useful as small datasets, will try to pre-training our RL network on these for it to gain experience with hyperparameter optimization before setting it loose on language model param optimization). (Lastly, hoping to also work out progressive learning, where the model gets really good at one subject, then gets better at the next subject, which is easy enough with math, but we can then reuse for progressive word conjugation mastery when making our network to be multilingual).

File: model.py (hunk 3 above)
Comment (gkielian) [reply model.py:163]: Good catch! : ) will update very soon.

File: data/modular_addition/README.md (hunk 2 above)
Comment (gkielian) [reply data/modular_addition/README.md]: Oh, this is for parameter reduction via tokenization optimization. 

The generator create exhaustively all addition examples of modulo 'x', then represents them in different bases.

The transformer training looks very different in each of the runs, and the training and validation losses take longer to settle depending on the number of tokens they need to learn to combine.

The number of tokens needed to represent a number is larger with lower bases (e.g. one token 'A' in hex, two tokens '12' in octal,  or four tokens '1010' in base 2). 

The more tokens, the larger the number of iterations we've observed for the network to converge, and the bigger the 'transition energy' into the final state as well. 

Will have graphs this Friday.

**Comment by gkielian** [comment]
Will try to refactor this after starting a few long-running data-augmentation scripts.  Retracting for the time being.
